mlkem.go

```diff
@@ -0,0 +1,361 @@
+//go:build !cmd_go_bootstrap && (cgo || goexperiment.ms_nocgo_opensslcrypto)
+
+package openssl
+
+import (
+	"errors"
+	"sync"
+	"unsafe"
+
+	"github.com/golang-fips/openssl/v2/internal/ossl"
+)
+
+const (
+	// sharedKeySizeMLKEM is the size of a shared key produced by ML-KEM.
+	sharedKeySizeMLKEM = 32
+
+	// seedSizeMLKEM is the size of a seed used to generate a decapsulation key.
+	seedSizeMLKEM = 64
+
+	// ciphertextSizeMLKEM768 is the size of a ciphertext produced by ML-KEM-768.
+	ciphertextSizeMLKEM768 = 1088
+
+	// encapsulationKeySizeMLKEM768 is the size of an ML-KEM-768 encapsulation key.
+	encapsulationKeySizeMLKEM768 = 1184
+
+	// ciphertextSizeMLKEM1024 is the size of a ciphertext produced by ML-KEM-1024.
+	ciphertextSizeMLKEM1024 = 1568
+
+	// encapsulationKeySizeMLKEM1024 is the size of an ML-KEM-1024 encapsulation key.
+	encapsulationKeySizeMLKEM1024 = 1568
+)
+
+// SupportsMLKEM768 returns true if ML-KEM-768 is supported on this platform.
+func SupportsMLKEM768() bool {
+	if vMajor >= 3 && vMinor >= 5 {
+		return supportsMLKEM768()
+	}
+	return false
+}
+
+// SupportsMLKEM1024 returns true if ML-KEM-1024 is supported on this platform.
+func SupportsMLKEM1024() bool {
+	if vMajor >= 3 && vMinor >= 5 {
+		return supportsMLKEM1024()
+	}
+	return false
+}
+
+var supportsMLKEM768 = sync.OnceValue(func() bool {
+	sig, _ := ossl.EVP_KEYMGMT_fetch(nil, _KeyTypeMLKEM768.ptr(), nil)
+	if sig != nil {
+		ossl.EVP_KEYMGMT_free(sig)
+		return true
+	}
+	return false
+})
+
+var supportsMLKEM1024 = sync.OnceValue(func() bool {
+	sig, _ := ossl.EVP_KEYMGMT_fetch(nil, _KeyTypeMLKEM1024.ptr(), nil)
+	if sig != nil {
+		ossl.EVP_KEYMGMT_free(sig)
+		return true
+	}
+	return false
+})
+
+// DecapsulationKeyMLKEM768 is the secret key used to decapsulate a shared key
+// from a ciphertext. It includes various precomputed values.
+type DecapsulationKeyMLKEM768 [seedSizeMLKEM]byte
+
+// GenerateKeyMLKEM768 generates a new decapsulation key, drawing random bytes from
+// the default crypto/rand source. The decapsulation key must be kept secret.
+func GenerateKeyMLKEM768() (DecapsulationKeyMLKEM768, error) {
+	var dk DecapsulationKeyMLKEM768
+	_, err := RandReader.Read(dk[:])
```

Comment:
This is still not using OpenSSL MLKEM API. I was expecting something like `EVP_PKEY_Q_keygen` here.

Reply:
updated PTAL